src/components/Table/EditColumnsButton.tsx

```diff
@@ -0,0 +1,93 @@
+import { Dispatch, SetStateAction, useCallback, useEffect, useMemo, useRef, useState } from "react";
+import { useMenuTrigger } from "react-aria";
+import { useMenuTriggerState } from "react-stately";
+import { CheckboxGroup, CheckboxGroupItemOption } from "src/inputs";
+import { Css } from "../../Css";
+import { useTestIds } from "../../utils";
+import { Button } from "../Button";
+import { isIconButton, isTextButton, OverlayTrigger, OverlayTriggerProps } from "../internal/OverlayTrigger";
+import { GridColumn, Kinded } from "./GridTable";
+import { GridColumnState } from "./useColumns";
+
+interface EditColumnsButtonProps<R extends Kinded, S>
+  extends Pick<OverlayTriggerProps, "trigger" | "placement" | "disabled" | "tooltip"> {
+  columns: GridColumn<R, S>[];
+  setColumns: Dispatch<SetStateAction<GridColumnState<R, S>>>;
+  title?: string;
+  // for storybook purposes
+  defaultOpen?: boolean;
+}
+
+export function EditColumnsButton<R extends Kinded, S = {}>(props: EditColumnsButtonProps<R, S>) {
+  const { defaultOpen, disabled, columns, setColumns, trigger, title } = props;
+  const state = useMenuTriggerState({ isOpen: defaultOpen });
+  const buttonRef = useRef<HTMLButtonElement>(null);
+  const { menuTriggerProps } = useMenuTrigger({ isDisabled: !!disabled }, state, buttonRef);
+  const tid = useTestIds(
+    props,
+    isTextButton(trigger) ? trigger.label : isIconButton(trigger) ? trigger.icon : trigger.name,
+  );
+
+  const editableColumns = useMemo(
+    () =>
+      columns.filter((column) => {
+        if (!column.canHide) return false;
+        if (!column.name || column.name.length === 0) {
+          console.warn("Column is missing 'name' property required by the Edit Columns button", column);
+          return false;
+        }
+        return true;
+      }),
+    [columns],
+  );
+
+  const selectedColumns: string[] = editableColumns
+    .map((column) => {
+      if (column.canHide && column.visible) return column.name;
+    })
+    .filter((column) => !!column) as string[];
+
+  const options: CheckboxGroupItemOption[] = editableColumns.map((column) => ({
+    label: column.name!,
+    value: column.name!,
+  }));
+
+  const [selectedValues, setSelectedValues] = useState<string[]>(selectedColumns ?? []);
```

Comment:
**question:** Can we accomplish this logic with introducing a second state variable for the selected columns?

Reply:
The selected values state handles the selected columns, what is the reason to use another state for it? I'm not clear

Reply:
The `columns` for `EditColumnsButton` are all columns for the table, is not the same as `useColumns` that handles `columns.visibleColumns`  for visible columns in the table, while setSelectedValues should handle all columns with canHide no matter if it is visible or not in the table